lib/sync_engine/2020_01_15/user_manager.rb

```diff
@@ -12,25 +12,28 @@ def change_pw(user, password, params)
 
         # We want to create a new session only if upgrading from a protocol version that does not
         # support sessions (i.e: 003 to 004).
-        if upgrading_protocol_version
-          handle_auth_response(user, params)
+        if upgrading_protocol_version && new_protocol_version == @user_class::SESSIONS_PROTOCOL_VERSION
+          handle_successful_authentication(user, params)
         else
           { user: user }
         end
       end
 
       private
 
-      def handle_auth_response(user, params)
-        session = Session.new(user_uuid: user.uuid, api_version: params[:api], user_agent: params[:user_agent])
+      def handle_successful_authentication(user, params)
+        # Users can be using the new API version but may not be on a protocol that support sessions.
+        if !user.supports_sessions?
+          return super(user, params)
+        end
+
+        session = user.sessions.new(api_version: params[:api], user_agent: params[:user_agent])
 
         unless session.save
           return { error: { message: 'Could not create a session.', status: 400 } }
         end
 
-        response = session.response_hash
```

Comment:
Oh I thought response_hash was a built-in method. In this case, it seems like the model is using inside information to dictate what a controller should return? This doesn't make sense imo. A model shouldn't prebuild controller response objects. That's the controller's responsibility. A model shouldn't know that the controller expects that the client expects a response with a nested `session` object.

Reply:
That is, https://github.com/standardnotes/syncing-server/pull/56/commits/d35f7d13b0693b17e6cf6b1d42ec781a7a2d4582#diff-a5520e44c2e541c1da8dae178b1b8996R65 shouldn't exist and should be moved into the controller instead.